library/alloc/src/boxed/thin.rs

```diff
@@ -0,0 +1,219 @@
+                let pointer = ptr::from_raw_parts_mut::<T>(value as *mut (), metadata);
+                &mut *pointer
+            };
+            self.ptr.drop::<T>(value);
+        }
+    }
+}
+
+#[unstable(feature = "thin_box", issue = "92791")]
+impl<T: ?Sized> ThinBox<T> {
+    fn meta(&self) -> <T as Pointee>::Metadata {
+        //  Safety:
+        //  -   NonNull and valid.
+        unsafe { *self.ptr.header() }
+    }
+
+    fn data(&self) -> *mut u8 {
+        self.ptr.value()
+    }
+}
+
+struct WithHeader<H>(NonNull<u8>, PhantomData<H>);
+
+impl<H> WithHeader<H> {
+    #[cfg(not(no_global_oom_handling))]
+    fn new<T>(header: H, value: T) -> WithHeader<H> {
+        let layout = Self::alloc_layout(mem::size_of::<T>(), mem::align_of::<T>());
+        let aligned_header_size = Self::aligned_header_size(layout.align());
+
+        unsafe {
+            let ptr = alloc::alloc(layout);
+
+            if ptr.is_null() {
+                alloc::handle_alloc_error(layout);
+            }
+            //  Safety:
+            //  -   The size is at least `aligned_header_size`.
+            let ptr = ptr.offset(aligned_header_size as isize) as *mut _;
+
+            let ptr = NonNull::new_unchecked(ptr);
+
+            let result = WithHeader(ptr, PhantomData);
+            ptr::write(result.header(), header);
+            ptr::write(result.value().cast(), value);
+
+            result
+        }
+    }
+
+    //  Safety:
+    //  -   Assumes that `value` can be dereferenced.
+    unsafe fn drop<T: ?Sized>(&self, value: &mut T) {
+        unsafe {
+            let layout = Self::alloc_layout(mem::size_of_val(value), mem::align_of_val(value));
+            let aligned_header_size = Self::aligned_header_size(layout.align());
+
+            ptr::drop_in_place::<T>(value as *mut T);
+            alloc::dealloc(self.0.as_ptr().offset(-(aligned_header_size as isize)), layout);
+        }
+    }
+
+    fn header(&self) -> *mut H {
+        //  Safety:
+        //  -   At least `size_of::<H>()` bytes are allocated ahead of the pointer.
+        unsafe { self.0.as_ptr().offset(-(Self::header_size() as isize)) as *mut H }
+    }
+
+    fn value(&self) -> *mut u8 {
+        self.0.as_ptr()
+    }
+
+    fn header_size() -> usize {
+        mem::size_of::<H>()
+    }
+
+    fn alloc_layout(value_size: usize, value_align: usize) -> Layout {
+        assert!(Self::header_size() <= usize::MAX / 2 + 1);
+
+        let alloc_align = cmp::max(mem::align_of::<H>(), value_align);
+        let alloc_size = Self::aligned_header_size(alloc_align) + value_size;
+
+        //  Safety:
+        //  -   `align` is not zero and a power of two, as guaranteed by `mem::align_of`.
+        //  -   The size can be rounded up to `align` without overflow, or `H` is way too big.
+        unsafe { Layout::from_size_align_unchecked(alloc_size, alloc_align) }
+    }
+
+    fn aligned_header_size(alloc_align: usize) -> usize {
+        cmp::max(Self::header_size(), alloc_align)
```

Comment:
This looks a bit suspicious, esp in conjunction with how alloc size is computed above.

I think this will work for currently existing types, but will fail for things where with metadata larger than the value (or header) alignment, but not aligned to value align.

For example, metadata type as `[u8; 17]`, and value type as `u64` (I know this is a bit silly, but we should probably not leave this bug until later). Then the alloc_align will be 8 (for u64), `align_of::<H>()` will be 1, and we'll end up storing the u64 at an unaligned address. (Sorry if this isn't explained well)

So, I think this should be aligning/rounding up here, rather than computing the max. I also like https://doc.rust-lang.org/nightly/std/alloc/struct.Layout.html#method.extend for this kind of thing, although perhaps it's annoying to integrate here without a performance overhead from the `Result`. You could use it to check the values in `alloc_layout` in debug builds, though.